File: utilities/collections/csharp/CollectionUtils.cs
namespace PolyglotUtilities.Collections;

public static class CollectionUtils
{
    public static List<T> Unique<T>(IEnumerable<T> items)
    {
        return items.Distinct().ToList();
    }

    public static Dictionary<TKey, List<T>> GroupBy<T, TKey>(
        IEnumerable<T> items,
        Func<T, TKey> keySelector) where TKey : notnull
    {
        var groups = new Dictionary<TKey, List<T>>();
        foreach (var item in items)
        {
            var key = keySelector(item);
            if (!groups.TryGetValue(key, out var values))
            {
                values = new List<T>();
                groups[key] = values;
            }

            values.Add(item);
        }

        return groups;
    }

    public static double Sum(IEnumerable<double> values)
    {
        return values.Sum();
    }

    public static double Average(IEnumerable<double> values)
    {
        var materialized = values.ToList();
        return materialized.Count == 0 ? 0 : materialized.Average();
    }
}

File: utilities/collections/typescript/collectionUtils.ts
export const unique = <T>(items: T[]): T[] => [...new Set(items)];

export const groupBy = <T, K extends keyof T>(items: T[], key: K): Record<string, T[]> => items.reduce<Record<string, T[]>>((groups, item) => {
  const group = String(item[key]);
  groups[group] = groups[group] || [];
  groups[group].push(item);
  return groups;
}, {});

export const sum = (values: number[]): number => values.reduce((total, value) => total + value, 0);

export const average = (values: number[]): number => values.length === 0 ? 0 : sum(values) / values.length;
